drivers/dma/dma_smartbond.c

```diff
@@ -0,0 +1,937 @@
+		LOG_ERR("Requested DMA channel should first be configured");
+		return -EINVAL;
+	}
+
+	if (size == 0) {
+		LOG_ERR("Min. transfer size is one");
+		return -EINVAL;
+	}
+
+	if (DMA_CTRL_REG_GET_FIELD(DMA_ON, regs->DMA_CTRL_REG)) {
+		LOG_ERR("Channel is busy, settings cannot be changed mid-transfer");
+		return -EBUSY;
+	}
+
+	if (src % data->channel_data[channel].bus_width) {
+		LOG_ERR("Source address is not bus width aligned");
+		return -EINVAL;
+	}
+
+	if (!dma_channel_src_addr_check_and_adjust(channel, &src)) {
+		return -EINVAL;
+	}
+
+	regs->DMA_A_START = src;
+
+	if (dst % data->channel_data[channel].bus_width) {
+		LOG_ERR("Destination address is not bus width aligned");
+		return -EINVAL;
+	}
+
+	if (!dma_channel_dst_addr_check_and_adjust(channel, &dst)) {
+		return -EINVAL;
+	}
+
+	regs->DMA_B_START = dst;
+
+	if (size % (data->channel_data[channel].burst_len *
+							data->channel_data[channel].bus_width)) {
+		LOG_ERR("Requested data size is not multiple of bus width");
+		return -EINVAL;
+	}
+
+	regs->DMA_LEN_REG = (size / data->channel_data[channel].bus_width) - 1;
+
+	/* Interrupt will be raised once all transfers are complete. */
+	regs->DMA_INT_REG = (size / data->channel_data[channel].bus_width) - 1;
+
+	return 0;
+}
+
+static int dma_smartbond_start(const struct device *dev, uint32_t channel)
+{
+	struct channel_regs *regs;
+	struct dma_smartbond_data *data = dev->data;
+
+	if (channel >= DMA_CHANNEL_COUNT) {
+		LOG_ERR("Inavlid DMA channel index");
+		return -EINVAL;
+	}
+	regs = DMA_CHN2REG(channel);
+
+	if (!data->channel_data[channel].is_dma_configured) {
+		LOG_ERR("Requested DMA channel should first be configured");
+		return -EINVAL;
+	}
+
+	/* Should return succss if the requested channel is already started. */
+	if (DMA_CTRL_REG_GET_FIELD(DMA_ON, regs->DMA_CTRL_REG)) {
+		return 0;
+	}
+
+	dma_smartbond_set_channel_status(channel, true);
+
+	return 0;
+}
+
+static int dma_smartbond_stop(const struct device *dev, uint32_t channel)
+{
+	struct channel_regs *regs;
+
+	if (channel >= DMA_CHANNEL_COUNT) {
+		LOG_ERR("Inavlid DMA channel index");
+		return -EINVAL;
+	}
+	regs = DMA_CHN2REG(channel);
+
+	/*
+	 * In normal mode DMA_ON is cleared automatically. However we need to clear
+	 * the corresponding register mask and disable NVIC if there is no other
+	 * channel in use.
+	 */
+	dma_smartbond_set_channel_status(channel, false);
+
+	return 0;
+}
+
+static int dma_smartbond_suspend(const struct device *dev, uint32_t channel)
+{
+	struct dma_smartbond_data *data = dev->data;
+
+	if (channel >= DMA_CHANNEL_COUNT) {
+		LOG_ERR("Inavlid DMA channel index");
+		return -EINVAL;
+	}
+
+	/*
+	 * Freezing the DMA engine is valid for memory-to-memory operations.
+	 * Valid memory locations are SYSRAM and/or PSRAM.
+	 */
+	if (data->channel_data[channel].dir != MEMORY_TO_MEMORY) {
+		LOG_ERR("Freezing DMA is valid for memory-to-memoy direction only.");
+		return -EINVAL;
+	}
```

Comment:
this check doesn't quite make sense if you can only freeze all channels, i.e. even if this channel is memory-to-memory there can be other active channels that are not and yet you'll still execute freeze.

Reply:
Correct. Removed the check completely. Freezing/Unfreezing only affects memory-to-memory channels.